src/callbacks/kfac/kfac.cpp

```diff
@@ -0,0 +1,482 @@
+    }
+    prof_region_end("kfac-update/local", prof_sync);
+
+#ifdef LBANN_NVPROF
+    prof_region_begin("kfac-update/local-barrier", prof_color, prof_sync);
+    CHECK_CUDA(cudaDeviceSynchronize());
+    comm->trainer_barrier();
+    prof_region_end("kfac-update/local-barrier", prof_sync);
+#endif // LBANN_NVPROF
+
+    // List-up buffers to synchronize.
+    std::vector<std::pair<size_t, El::AbstractMatrix<DataType>*>> buffers;
+    size_t global_buffer_size = 0;
+    for(auto& block : m_blocks)
+      for(auto L : block->get_local_kronecker_buffers()) {
+        const size_t rank = block->get_inverse_proc_rank();
+        buffers.emplace_back(rank, L);
+        assert(L->Width() == 1);
+        global_buffer_size += L->Height();
+      }
+
+    // Perform reduce-scatter.
+    prof_region_begin("kfac-update/reduce-scatter", prof_color, prof_sync);
+    const auto reduce_scatter_mode = kfac_reduce_scatter_mode::ALLREDUCE;
+    El::Matrix<DataType, Device>& global_buffer =
+        get_workspace_matrix(
+            "reduce_scatter_send_buffer",
+            kfac_util::is_reduce_scatter_buffer_required(reduce_scatter_mode) ? global_buffer_size : 0,
+            1);
+    kfac_util::reduce_scatter_blocks(
+        buffers, global_buffer, comm, reduce_scatter_mode);
+    prof_region_end("kfac-update/reduce-scatter", prof_sync);
+
+#ifdef LBANN_NVPROF
+    prof_region_begin("kfac-update/reduce-scatter-barrier", prof_color, prof_sync);
+    CHECK_CUDA(cudaDeviceSynchronize());
+    comm->trainer_barrier();
+    prof_region_end("kfac-update/reduce-scatter-barrier", prof_sync);
+#endif // LBANN_NVPROF
+
+    prof_region_begin("kfac-update/average", prof_color, prof_sync);
+    for(auto& block : m_blocks) {
+      prof_region_begin(("kfac-update/average/" + block->get_name()).c_str(), prof_color, prof_sync);
+      block->update_kronecker_average(
+          comm,
+          m_kronecker_decay,
+          m_print_matrix, m_print_matrix_summary);
+      prof_region_end(("kfac-update/average/" + block->get_name()).c_str(), prof_sync);
+    }
+    prof_region_end("kfac-update/average", prof_sync);
+
+    prof_region_end("kfac-update", prof_sync);
+  }
+
+  // Step 2: Model-parallel inverse computation
+  prof_region_begin("kfac-inverse", prof_color, prof_sync);
+  for(auto& block : m_blocks) {
+    if(!is_kronecker_update_required || (size_t) comm->get_rank_in_trainer() != block->get_inverse_proc_rank())
+      continue;
+
+    prof_region_begin(("kfac-inverse/" + block->get_name()).c_str(), prof_color, prof_sync);
+    // TODO: Add kfac_block::is_bn?
+    const bool is_bn = dynamic_cast<kfac_block_bn<Device>*>(block.get()) != nullptr;
+    block->update_kronecker_inverse(
+        comm, m_use_pi,
+        is_bn ? m_damping_bn_act : m_damping_act,
+        is_bn ? m_damping_bn_err : m_damping_err,
+        m_learning_rate_factor,
+        m_print_matrix, m_print_matrix_summary,
+        m_print_time);
```

Comment:
If I understand this function correctly, it inverts the Fisher matrix and applies it to precondition the gradient. This is only called once every `m_update_interval` steps. Would it be better to precondition the gradients at every step, even with a stale Fisher matrix? To do this, we should split this function into two parts.

Reply:
I agree that that will improve the performance (I was mainly using `m_update_interval=1` so I haven't investigated that effect in detail).